From 5fc93a33144fca1c27bbfe6fbeb4ebfb43b4d692 Mon Sep 17 00:00:00 2001
From: Sunil Goutham <sgoutham@marvell.com>
Date: Fri, 21 Aug 2020 18:35:50 +0530
Subject: [PATCH 626/767] octeontx2-pf: Fix receive buffer size calculation

commit 1634ec59fefebed6314484b7223a05956edd26e6 from
git@git.assembla.com:cavium/WindRiver.linux.git

Not considering L2 header while calculating receive buffer
size resulting in pkt spilling over to next buffer. Since
we no longer support multi-segmented receive pkts while
processing CQE_RX these pkts are treated as errors and dropped.
This patch fixes the issue to makesure pkt fits in single segment.

Change-Id: I6dd49e9bc66aa4a4380b5736afd083d9aab982af
Signed-off-by: Sunil Goutham <sgoutham@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/34367
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
---
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
index 7dab96d75ff0..35e0283a7f47 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
@@ -1326,6 +1326,16 @@ static int otx2_init_hw_resources(struct otx2_nic *pf)
 	pf->rbsize = RCV_FRAG_LEN(OTX2_HW_TIMESTAMP_LEN + pf->netdev->mtu +
 				  pf->addl_mtu + pf->xtra_hdr);
 
+	/* Get the size of receive buffers to allocate.
+	 *
+	 * Depending interface mode ie timestamp and/or Higig2 enabled
+	 * and presence of DSA tags in L2 header, the packet size would
+	 * vary, hence consider all while calculating receive buffer size.
+	 */
+	pf->rbsize = pf->netdev->mtu + OTX2_ETH_HLEN;
+	pf->rbsize += OTX2_HW_TIMESTAMP_LEN + pf->addl_mtu + pf->xtra_hdr;
+	pf->rbsize = RCV_FRAG_LEN(pf->rbsize);
+
 	otx2_mbox_lock(mbox);
 	/* NPA init */
 	err = otx2_config_npa(pf);
-- 
2.31.1

